lib/std/channels.nim

```diff
@@ -0,0 +1,511 @@
+
+  release(chan.headLock)
+  signal(chan.notFullCond)
+  result = true
+
+proc channelCloseMpmc(chan: ChannelRaw): bool =
+  # Unsynchronized
+
+  if chan.isClosed:
+    # ChannelRaw already closed
+    return false
+
+  store(chan.closed, true, moRelaxed)
+  result = true
+
+proc channelOpenMpmc(chan: ChannelRaw): bool =
+  # Unsynchronized
+
+  if not chan.isClosed:
+    # ChannelRaw already open
+    return false
+
+  store(chan.closed, false, moRelaxed)
+  result = true
+
+# Public API
+# ----------------------------------------------------------------------------------
+
+type
+  Channel*[T] = object ## Typed channels
+    d: ChannelRaw
+
+proc `=destroy`*[T](c: var Channel[T]) =
+  if c.d != nil:
+    if load(c.d.atomicCounter, moAcquire) == 0:
+      if c.d.buffer != nil:
+        freeChannel(c.d)
+    else:
+      atomicDec(c.d.atomicCounter)
+
+proc `=`*[T](dest: var Channel[T], src: Channel[T]) =
+  ## Shares `Channel` by reference counting.
+  if src.d != nil:
+    atomicInc(src.d.atomicCounter)
+
+  if dest.d != nil:
+    `=destroy`(dest)
+  dest.d = src.d
+
+proc channelSend[T](chan: Channel[T], data: sink T, size: int, nonBlocking: bool): bool {.inline.} =
+  ## Send item to the channel (FIFO queue)
+  ## (Insert at last)
+  sendMpmc(chan.d, data.unsafeAddr, size, nonBlocking)
+
+proc channelReceive[T](chan: Channel[T], data: ptr T, size: int, nonBlocking: bool): bool {.inline.} =
+  ## Receive an item from the channel
+  ## (Remove the first item)
+  recvMpmc(chan.d, data, size, nonBlocking)
+
+func trySend*[T](c: Channel[T], src: var Isolated[T]): bool {.inline.} =
+  ## Sends item to the channel(non blocking).
+  var data = src.extract
+  result = channelSend(c, data, sizeof(data), true)
+  if result:
+    wasMoved(data)
+
+template trySend*[T](c: Channel[T], src: T): bool =
+  ## Helper templates for `trySend`.
+  trySend(c, isolate(src))
+
+func tryRecv*[T](c: Channel[T], dst: var T): bool {.inline.} =
+  ## Receives item from the channel(non blocking).
+  channelReceive(c, dst.addr, sizeof(dst), true)
+
+func send*[T](c: Channel[T], src: sink Isolated[T]) {.inline.} =
+  ## Sends item to the channel(blocking).
+  var data = src.extract
+  discard channelSend(c, data, sizeof(data), false)
+  wasMoved(data)
+
+template send*[T](c: var Channel[T]; src: T) =
+  ## Helper templates for `send`.
+  send(c, isolate(src))
+
+func recv*[T](c: Channel[T], dst: var T) {.inline.} =
+  ## Receives item from the channel(blocking).
+  discard channelReceive(c, dst.addr, sizeof(dst), false)
+
+func recvIso*[T](c: Channel[T]): Isolated[T] {.inline.} =
+  var dst: T
+  discard channelReceive(c, dst.addr, sizeof(dst), false)
+  result = isolate(dst)
+
+func open*[T](c: Channel[T]): bool {.inline.} =
+  result = c.d.channelOpenMpmc()
+
+func close*[T](c: Channel[T]): bool {.inline.} =
+  result = c.d.channelCloseMpmc()
+
+func peek*[T](c: Channel[T]): int {.inline.} = peek(c.d)
+
+proc newChannel*[T](elements = 30): Channel[T] =
```

Comment:
* document that `elements = 1` will cause channel to be unbuffered
* magic number 30 sounds weird; how about making the default unbuffered instead (and using an impossible value, say -1, to denote that the channel is unbufferred)

Reply:
make sense